File: leetcode/pacific_atlantic_water_flow/pacific_atlantic_water_flow.cpp
/*
 * Time Complexity: O(m * n), where m = number of rows and n = number of columns in heights.
 * Each cell is processed at most twice (once for Pacific, once for Atlantic), and each DFS/BFS traversal
 * visits neighbors in O(4) per cell. So overall O(m * n).
 *
 * Space Complexity: O(m * n) for the auxiliary matrix `aux` and the recursion/stack used for DFS traversal.
 * The stack can, in the worst case, hold all cells, so O(m * n).
 */

#include <vector>
#include <stack>
#include <string>

using std::vector;
using std::stack;
using std::pair;
using std::string;

class Solution {
private:
    bool isOutOfBounds(const int& i, const int& j, const int& m, const int& n) {
        return i < 0 || j < 0 || i >= m || j >= n;
    }

public:
    vector<vector<int>> pacificAtlantic(vector<vector<int>>& heights) {
        int m = heights.size();
        int n = heights[0].size();
        int dir [4][2] = {{-1,0}, {0,1}, {1,0}, {0,-1}};

        vector<vector<string>> aux(m, vector<string>(n, ""));
        stack<pair<int, int>> coord;

        for(int j = 0; j < n; j++) coord.push({0,j});
        for(int i = 0; i < m; i++) coord.push({i,0});

        while(!coord.empty()) {
            int i = coord.top().first;
            int j = coord.top().second;
            coord.pop();

            if(aux[i][j] != "") continue;
            aux[i][j] = "P";

            for(int k = 0; k < 4; k++) {
                int x = i + dir[k][0];
                int y = j + dir[k][1];

                if(!isOutOfBounds(x, y, m, n) && aux[x][y] != "P" && heights[x][y] >= heights[i][j]) {
                    coord.push({x,y});
                }
            }
        }

        for(int j = 0; j < n; j++) coord.push({m-1,j});
        for(int i = 0; i < m; i++) coord.push({i,n-1});

        while(!coord.empty()) {
            int i = coord.top().first;
            int j = coord.top().second;
            coord.pop();

            if(aux[i][j] == "A" || aux[i][j] == "PA") continue;
            aux[i][j] += "A";

            for(int k = 0; k < 4; k++) {
                int x = i + dir[k][0];
                int y = j + dir[k][1];

                if(!isOutOfBounds(x, y, m, n) && aux[x][y] != "A" && aux[x][y] != "PA" && heights[x][y] >= heights[i][j]) {
                    coord.push({x,y});
                }
            }
        }

        vector<vector<int>> ans;
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++) {
                if(aux[i][j] == "PA") ans.push_back({i,j});
            }
        }

        return ans;
    }
};

File: leetcode/pacific_atlantic_water_flow/pacific_atlantic_water_flow.java
package pacific_atlantic_water_flow;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Deque;
import java.util.List;

class Solution {
    private boolean isOutOfBounds(int i, int j, int m, int n) {
        return i < 0 || j < 0 || i >= m || j >= n;
    }

    public List<List<Integer>> pacificAtlantic(int[][] heights) {
        int m = heights.length;
        int n = heights[0].length;
        int[][] dir = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};

        String[][] aux = new String[m][n];
        for (int i = 0; i < m; i++) Arrays.fill(aux[i], "");

        Deque<int[]> coord = new ArrayDeque<>();

        for (int j = 0; j < n; j++) coord.push(new int[]{0, j});
        for (int i = 0; i < m; i++) coord.push(new int[]{i, 0});

        while (!coord.isEmpty()) {
            int[] top = coord.pop();
            int i = top[0], j = top[1];

            if (!aux[i][j].equals("")) continue;
            aux[i][j] = "P";

            for (int[] d : dir) {
                int x = i + d[0];
                int y = j + d[1];

                if (!isOutOfBounds(x, y, m, n) && !aux[x][y].equals("P") && heights[x][y] >= heights[i][j]) {
                    coord.push(new int[]{x, y});
                }
            }
        }

        for (int j = 0; j < n; j++) coord.push(new int[]{m - 1, j});
        for (int i = 0; i < m; i++) coord.push(new int[]{i, n - 1});

        while (!coord.isEmpty()) {
            int[] top = coord.pop();
            int i = top[0], j = top[1];

            if (aux[i][j].equals("A") || aux[i][j].equals("PA")) continue;
            aux[i][j] += "A";

            for (int[] d : dir) {
                int x = i + d[0];
                int y = j + d[1];

                if (!isOutOfBounds(x, y, m, n) && !aux[x][y].equals("A") && !aux[x][y].equals("PA") && heights[x][y] >= heights[i][j]) {
                    coord.push(new int[]{x, y});
                }
            }
        }

        List<List<Integer>> ans = new ArrayList<>();
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (aux[i][j].equals("PA")) {
                    ans.add(Arrays.asList(i, j));
                }
            }
        }

        return ans;
    }
}
